frontend/src/metabase/query_builder/containers/QueryBuilder.unit.spec.tsx

```diff
@@ -211,219 +230,290 @@ describe("QueryBuilder", () => {
+        userEvent.click(await screen.findByText("Done"));
+        expect(await screen.findByText("Count")).toBeInTheDocument();
 
         const mockEvent = callMockEvent(mockEventListener, "beforeunload");
         expect(mockEvent.preventDefault).not.toHaveBeenCalled();
-        expect(mockEvent.returnValue).toBe(undefined);
+        expect(mockEvent.returnValue).not.toEqual(
+          BEFORE_UNLOAD_UNSAVED_MESSAGE,
+        );
       });
 
-      it("should not trigger beforeunload event when model metadata is unedited", async () => {
+      it("should not trigger beforeunload event when query is unedited", async () => {
         const { mockEventListener } = await setup({
-          card: TEST_MODEL_CARD,
-          dataset: TEST_MODEL_DATASET,
-          initialRoute: `/model/${TEST_MODEL_CARD.id}/metadata`,
+          card: TEST_STRUCTURED_CARD,
         });
 
         const mockEvent = callMockEvent(mockEventListener, "beforeunload");
         expect(mockEvent.preventDefault).not.toHaveBeenCalled();
-        expect(mockEvent.returnValue).toBe(undefined);
+        expect(mockEvent.returnValue).not.toEqual(
+          BEFORE_UNLOAD_UNSAVED_MESSAGE,
+        );
       });
     });
   });
 
-  describe("beforeunload events in native queries", () => {
-    afterEach(() => {
-      jest.restoreAllMocks();
-    });
-
-    it("should trigger beforeunload event when leaving edited question", async () => {
-      const { mockEventListener } = await setup({
+  describe("downloading results", () => {
+    it("should allow downloading results for a native query", async () => {
+      const mockDownloadEndpoint = fetchMock.post(
+        `/api/card/${TEST_NATIVE_CARD.id}/query/csv`,
+        {},
+      );
+      await setup({
         card: TEST_NATIVE_CARD,
+        dataset: TEST_NATIVE_CARD_DATASET,
       });
 
       const inputArea = within(
         screen.getByTestId("mock-native-query-editor"),
       ).getByRole("textbox");
 
-      userEvent.click(inputArea);
-      userEvent.type(inputArea, "0");
-
-      userEvent.tab();
+      expect(inputArea).toHaveValue("SELECT 1");
 
-      // default native query is `SELECT 1`
-      expect(inputArea).toHaveValue("SELECT 10");
+      userEvent.click(screen.getByTestId("download-button"));
+      userEvent.click(await screen.findByRole("button", { name: ".csv" }));
 
-      const mockEvent = callMockEvent(mockEventListener, "beforeunload");
-      expect(mockEvent.preventDefault).toHaveBeenCalled();
-      expect(mockEvent.returnValue).toEqual(BEFORE_UNLOAD_UNSAVED_MESSAGE);
+      expect(mockDownloadEndpoint.called()).toBe(true);
     });
 
-    it("should not trigger beforeunload event when user tries to leave an ad-hoc native query", async () => {
-      const { mockEventListener } = await setup({
-        card: TEST_UNSAVED_NATIVE_CARD,
+    it("should allow downloading results for a native query using the current result even the query has changed but not rerun (metabase#28834)", async () => {
```

Comment:
As stated in the comment in `e2e/test/scenarios/sharing/downloads/reproductions/28834-modified-native-question.cy.spec.js`, I have tried to test unsaved (ad-hoc) queries in unit test, but it's too hard since I need to wire the mock `NativeQueryEditor` properly, so I test that scenario in Cypress instead. Also, having at least one Cypress test would make sure the bug is fixed properly.